keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,532 @@
+```
+
+And
+
+```go
+klog.ErrorS(err, "Failed to update pod status")
+```
+
+That would result in log
+```
+E1025 00:15:15.525108       1 controller_utils.go:114] "Failed to update pod status" err="timeout"
+```
+
+### References to Kubernetes objects
+
+To address problem of inconsistency in referencing of Kubernetes objects, new klog methods will include additional logic for handling their formatting. Kubernetes objects will be recognized by checking for `ObjectMeta` structure.
+
+Namespaced objects will be represented by joining their name and namespace with `/` character. Pattern: `<namespace>/<name>`, example for pod: `kube-system/kubedns`
+
+Non-namespaced objects will be represented by their name. Pattern: `<name>`, example for node: `cluster1-vm-72x33b8p-34jz`
+
+Example:
+```go
+pod := corev1.Pod{Name: "kubedns", Namespace: "kube-system", ...}
+klog.InfoS("Pod status updated", "pod", pod, "status", "ready")
+```
+That would result in log
+```
+I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+```
+Selecting key under which put objects will be left to caller as defining log schema will need more discussion and is outside of scope of this KEP.
+
+### Selecting most important logs
+
+As migration to new message structure will be done manually we will be focusing on logs that have the greatest impact for log querying and processing. We will be focusing on log messages Proposed plan of measurements:
+
+* Using log data from single run of kubernetes [gce-master-scale-performance](https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-master-scale-performance) tests
+* Based on logs generated on master node
+* Aggregate from components:
+  * kube-controller-manager
+  * kube-scheduler
+  * kube-apiserver
+  * kubelet
+* Only for log lines generated by klog (e.g. prefix `I0129 03:26:19.857798       1 garbagecollector.go:517]`)
+* Identified by code line that generated log (e.g. `garbagecollector.go:517`)
+
+We assume that exact details of taking measurements can be improved, but overall methodology is solid. With those criteria we concluded that covering 99.9% of logs on master node are generated by 22 log messages. Exact list of those logs is provided in the detail design section.
+
+#### JSON output format
+
+Introduction of new methods to klog library will make identification of different components of the log message much easier.
+With klog v2 we can take further advantage of this fact and add an option to produce structured logs in JSON format.
+
+Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Proposed default fields:
+* `ts` - timestamp as Unix time (required, `float64`)
+* `v` - verbosity (required, `int`, default 4)
```

Comment:
Also: do we need to denote info logs from error logs?

Reply:
> Why is there a default for `v`? Every log line has one, so you don't need a default. and wouldn't the default be 0?

Yes, it should be 0.

Reply:
> Also: do we need to denote info logs from error logs?

For now we will stick with only having optional `error` key. As it's for JSON logs it will be easy for filter error logs by using `jq`.

It could be a case that `error` field can be used in info logs. We will information into user guide about default fields having special meaning and discouraging from using them as keys.